File: Python/count_smileys.py
"""
|--------------------------------------------------------------------------
| Count the smiley faces!
|--------------------------------------------------------------------------
|
"""

def count_smileys(arr):
   smileys = [":)", ";)", ":-)", ";-)", ";~)", ":~)", ":D", ";D", ":-D", ":~D", ";-D", ";~D"]
   count = 0
   for i in range(len(arr)):
      for j in range(len(smileys)):
         if arr[i] == smileys[j]:
            count += 1
   return count


from re import findall, match
def count_smileys(arr):
   return len(list(findall(r"[:;][-~]?[)D]", " ".join(arr))))


def count_smileys(arr):
   return sum(1 for s in arr if match(r'\A[:;][-~]?[)D]\Z',s))


def count_smileys(arr):
   return len(list(filter(lambda x: x in [':D',':~D',':-D',';D',';~D',';-D',':)',':~)',':-)',';)',';~)',';-)'],arr)))


def count_smileys(arr):
   smiles = set([a+b+c for a in ":;" for b in ['','-', '~'] for c in ")D"])
   return len([1 for s in arr if s in smiles])


print(count_smileys([]), 0)
print(count_smileys([':D',':~)',';~D',':)']), 4)
print(count_smileys([':)',':(',':D',':O',':;']), 2)
print(count_smileys([';]', ':[', ';*', ':$', ';-D']), 1)

File: JavaScript/6js 1/countSmileys.js
/*
|--------------------------------------------------------------------------
| Count the smiley faces!
|--------------------------------------------------------------------------
|
*/

function countSmileys(arr) {
   let res = 0
   for (let i = 0; i < arr.length; i++) {
      let x = arr[i]
      if (x[0] == ':' || x[0] == ';') {
         if (x[x.length - 1] == ')' || x[x.length - 1] == 'D') {
            if (x.length == 3) {
               if (x[1] == '-' || x[1] == '~') {
                  res += 1
               }
            } else if (x.length == 2) {
               res += 1
            }
         }
      }
   }
   return res
}


function countSmileys(arr) {
   return arr.filter(x => /^[:;][-~]?[)D]$/.test(x)).length;
}


function countSmileys(arr) {
   let smileys = [":)", ";)", ":-)", ";-)", ";~)", ":~)", ":D", ";D", ":-D", ":~D", ";-D", ";~D"];
   let count = 0;
   for (let i = 0; i < arr.length; i++) {
      for (let j = 0; j < smileys.length; j++) {
         if (arr[i] === smileys[j]) {
            count++;
         }
      }
   }
   return count;
}


function countSmileys(arr) {
   return arr.reduce((n, s) => /^[:;][-~]?[)D]$/.test(s) ? n + 1 : n, 0)
}


console.log(countSmileys([]), 0);
console.log(countSmileys([':D', ':x)', ';~D', ':)']), 3);
console.log(countSmileys([':)', ':(', ':D', ':O', ':;']), 2);
console.log(countSmileys([';]', ':[', ';*', ':$', ';-D']), 1);